File: problems/3/Solution.swift
class Solution {
    func lengthOfLongestSubstring(_ s: String) -> Int {
        var longestLength = 0
        var map = [Character: Int]()
        var i = 0, j = 0
        
        for c in s.characters {
            if let next = map[c] {
                if j - i > longestLength {
                    longestLength = j - i
                }
                if next + 1 > i {
                    i = next + 1
                }
            }
            map[c] = j
            j += 1
        }
        
        if s.characters.count - i > longestLength {
            longestLength = s.characters.count - i
        }
        return longestLength
    }
}

File: problems/3/Solution.cpp
// 3. Longest Substring Without Repeating Characters

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int n = s.length();
        int left = 0;
        int right = 0;
        unordered_map<char, int> hashmap;   // character, right position
        int answer = 0;
        while (right < n) {
            if (hashmap.count(s[right]) > 0 && hashmap[s[right]] >= left) {
                left = hashmap[s[right]] + 1;
            }
            answer = std::max(answer, right - left + 1);
            hashmap[s[right]] = right;
            right++;
        }
        return answer;
    }
};
